drivers: firmware: TISCI driver support

Kronosblaster opened the pull request:
Fixed bugs in secureproxy mailbox driver
Added TISCI driver for supported devices using the binding ti,k2g-sci. 
Added TISCI AM2434 EVM
Added TISCI test sample

**API test:**
The test sample is present at /tests/drivers/tisci
## Test Sequence

1. **Version Information**
   - Retrieves TISCI firmware version and ABI compatibility info
   - Validates protocol communication basics
   - Reports ABI versions and firmware description

2. **Clock Management**
   - Tests clock ID 152 (main UART 1 clock)
   - Sequence:
     1. Reads initial frequency
     2. Sets frequency to 96MHz (min/target/max all set to 96MHz)
     3. Verifies frequency change by reading back

3. **Power Domain Control**
   - Tests device ID 0 (ADC)
   - Operations:
     1. Power-on sequence using `tisci_cmd_get_device`
     2. Reads device state (clock count, resets, power/clock states)
     3. Power-off sequence using `tisci_cmd_put_device`
     4. Verifies state transition

TISCI documentation of messages is given [here](https://software-dl.ti.com/tisci/esd/22_01_02/2_tisci_msgs/general/core.html)

vaishnavachath commented:
@Kronosblaster Please provide proper commit message/details to help a reviewer to understand what this is.

vaishnavachath commented on drivers/firmware/ti_sci/Kconfig:
```diff
@@ -0,0 +1,18 @@
+	  Init priority for the TISCI driver.
+
+config TISCI_RESPONSE_TIMEOUT_MS
+	int "TISCI response timeout in milliseconds"
```
Why is this configurable? does this change across devices?

Kronosblaster replied:
Added as a way to support devices that might have different response timings. Hardcoding the value also works for now.

vaishnavachath replied:
why is this needed? and why would this change across device families if it is already of the order of 100ms?

Kronosblaster replied:
As of testing on AM243x EVM it can be reliably pushed to a minimum of 2ms. Was not sure what the response timings would look like on other devices. Would it be better to have it as a define?

nordicjm commented on drivers/firmware/Kconfig:
```diff
@@ -15,4 +15,14 @@ source "drivers/firmware/nrf_ironside/Kconfig"
 source "drivers/firmware/scmi/Kconfig"
 # zephyr-keep-sorted-stop
 
+config TISCI
+	bool "TISCI Firmware driver"
+	select MBOX
+	help
+	  TISCI firmware driver is a frontend interface
+	  to TI System Controller interface firmware through
+	  secureproxy mailbox.
+
+source "drivers/firmware/ti_sci/Kconfig"
```
move this config to that file, put source file in list, correctly sorted, add correct depends on to this symbol

nordicjm commented on drivers/firmware/ti_sci/Kconfig:
```diff
@@ -0,0 +1,18 @@
+	int "TISCI response timeout in milliseconds"
+	default 100
+	help
+		Timeout in milliseconds to wait for a response from the TISCI firmware.
```
Kconfig help text indent is 1x tab followed by 2x spaces

Kronosblaster committed:
drivers: firmware: TISCI driver support

Added TISCI driver for supported devices using the binding ti,k2g-sci.
This is used to communicate via the secury proxy channel for clock,
resource and power domain management.
Refer: https://software-dl.ti.com/tisci/esd/latest/2_tisci_msgs/general/TISCI_header.html

Signed-off-by: Dave Joseph <d-joseph@ti.com>

glneo commented on drivers/firmware/ti_sci/ti_sci.h:
```diff
@@ -0,0 +1,1598 @@
+
+/**
+ * @file
+ * @brief Public APIs for the Device Multiplexer driver
```
??

natto1784 commented on drivers/firmware/ti_sci/ti_sci.h:
```diff
@@ -0,0 +1,1598 @@
+#define TI_SCI_FLAG_RESP_GENERIC_ACK       TI_SCI_MSG_FLAG(1)
+	/* Additional Flags */
+	uint32_t flags;
+} __packed;
```
Definition of the variable `__packed` multiple time throughout the header. Can use ` __attribute__((packed))` to get a packed struct declaration.

natto1784 replied:
there is a macro definition in `tisci_protocol.h` for `__packed` but none in the above header

Kronosblaster replied:
Will be added

natto1784 commented on drivers/firmware/ti_sci/ti_sci.c:
```diff
@@ -0,0 +1,1547 @@
+
+	ret = mbox_register_callback_dt(&config->mbox_rx, callback, NULL);
+	if (ret < 0) {
+		printk("Could not register callback (%d)\n", ret);
```
Minor nitpick: can use `LOG_*` macros here similar to other functions

natto1784 commented on drivers/firmware/ti_sci/ti_sci.c:
```diff
@@ -0,0 +1,1547 @@
+
+	ret = mbox_set_enabled_dt(&config->mbox_rx, true);
+	if (ret < 0) {
+		printk("Could not enable RX channel (%d)\n", ret);
```
same here

vaishnavachath commented on drivers/firmware/ti_sci/ti_sci.c:
```diff
@@ -0,0 +1,1547 @@
+	xfer = ti_sci_setup_one_xfer(dev, TI_SCI_MSG_SET_DEVICE_RESETS,
+				     TI_SCI_FLAG_REQ_ACK_ON_PROCESSED, (uint32_t *)&req,
+				     sizeof(req), sizeof(*resp));
+	if (!xfer) {
```
why are these checks again inconsistent? on all other instances you don't have this.

Kronosblaster commented on drivers/firmware/ti_sci/ti_sci.c:
```diff
@@ -0,0 +1,1547 @@
+	resp = (struct ti_sci_msg_resp_get_clock_parent *)xfer->rx_message.buf;
+
+	if (!resp) {
+		return -ENOMEM;
```
The rationale was that the required response is not available in memory. Will have it changed to EINVAL

Kronosblaster commented on drivers/firmware/ti_sci/ti_sci.c:
```diff
@@ -0,0 +1,1547 @@
+	int ret = 0;
+
+	if (!dev) {
+		return -EINVAL;
```
Added checks

glneo commented on drivers/firmware/ti_sci/ti_sci.c:
```diff
@@ -0,0 +1,1729 @@
+
+	/* Get response if requested */
+	if (xfer->rx_len) {
+		ret = ti_sci_get_response(dev, xfer);
```
`ret` is assigned here but never used, we either `return 0` or assign again to 'ret' before the one spot it could be returned.

glneo commented on drivers/firmware/ti_sci/ti_sci.c:
```diff
@@ -0,0 +1,1729 @@
+
+	req.dev_id = dev_id;
+	req.clk_id = clk_id;
+	ret = ti_sci_do_xfer(dev, xfer);
```
You don't check `ret` here, if the transfer failed, you should not continue to use the returned `resp` values.

glneo commented on drivers/firmware/ti_sci/ti_sci.c:
```diff
@@ -0,0 +1,1660 @@
+
+	/* Get response if requested */
+	if (xfer->rx_message.size) {
+		return ti_sci_get_response(dev, xfer);
```
Are you sure you want to always `return` here? Seems this would cause the `ti_sci_is_response_ack()` check below to be skipped in cases where we return data (which needs the ACK check also). Maybe moving the `ti_sci_is_response_ack()` above this to line 164 would be the right solution.

glneo commented on drivers/firmware/ti_sci/ti_sci.c:
```diff
@@ -0,0 +1,1660 @@
+	struct ti_sci_xfer *xfer;
+	int ret;
+
+	if (!dev) {
```
You don't do anything with `dev` in this function, it is only used by `ti_sci_setup_one_xfer()`, so if you do this `dev` check inside the setup function, you can remove this check from a whole bunch of spots.

glneo commented on drivers/firmware/ti_sci/ti_sci.c:
```diff
@@ -0,0 +1,1660 @@
+
+	xfer = ti_sci_setup_one_xfer(dev, TI_SCI_MSG_SET_CLOCK_FREQ,
+				     TI_SCI_FLAG_REQ_ACK_ON_PROCESSED, (uint32_t *)&req,
+				     sizeof(req), NULL, 0);
```
You should always check if the xfer was setup correctly, otherwise you might send a NULL ptr to `ti_sci_do_xfer()`. Plus it is more consistent that way.

glneo commented on drivers/firmware/ti_sci/ti_sci.c:
```diff
@@ -0,0 +1,1660 @@
+
+	xfer = ti_sci_setup_one_xfer(dev, TI_SCI_MSG_SET_CLOCK_STATE,
+				     flags | TI_SCI_FLAG_REQ_ACK_ON_PROCESSED, (uint32_t *)&req,
+				     sizeof(req), NULL, 0);
```
You should still have a response (not `NULL, 0`), even if it is just a `struct ti_sci_msg_hdr hdr`, that way you do all the normal response checks in `ti_sci_do_xfer()`, like sequence.

glneo commented on drivers/firmware/tisci/tisci.c:
```diff
@@ -0,0 +1,1542 @@
+	req.target_freq_hz = target_freq;
+	req.max_freq_hz = max_freq;
+
+	ret = tisci_do_xfer(dev, xfer);
```
No response check and error message here?

Kronosblaster replied:
I'll log the error. Isnt the response check in tisci_do_xfer() enough?

glneo commented on drivers/firmware/tisci/tisci.c:
```diff
@@ -0,0 +1,1599 @@
+	xfer = tisci_setup_one_xfer(dev, TISCI_MSG_GET_CLOCK_STATE, 0, &req, sizeof(req), &resp,
+				    sizeof(resp));
+	if (!xfer) {
+		LOG_ERR("%s:Failed to setup transfer", __func__);
```
I'd drop printing of the function name __func__ here, that isn't really useful and could be done as part of the LOG_ERR macro if we really wanted it for trace or similar.

glneo commented on drivers/firmware/tisci/tisci.c:
```diff
@@ -0,0 +1,1599 @@
+	req.dev_id = dev_id;
+	req.clk_id = clk_id;
+	ret = tisci_do_xfer(dev, xfer);
+
```
No newline between a function and the check for its return value. Remove that newline in the couple other spots below also.

glneo commented on drivers/firmware/tisci/tisci.c:
```diff
@@ -0,0 +1,1618 @@
+		return ret;
+	}
+
+	if ((resp.hdr.flags & TISCI_FLAG_RESP_GENERIC_ACK) != TISCI_FLAG_RESP_GENERIC_ACK) {
```
The response check is already done as part of `tisci_do_xfer()`, you can drop this check. Same below.

natto1784 commented on drivers/firmware/tisci/tisci.c:
```diff
@@ -0,0 +1,1599 @@
+	return 0;
+}
+
+static int tisci_init(const struct device *dev)
```
minor suggestion: can move `tisci_init` to the end of file, just above device init

natto1784 commented on drivers/firmware/tisci/tisci.c:
```diff
@@ -0,0 +1,1599 @@
+
+/* Received message container */
+static struct rx_msg rx_message;
+static uint8_t rx_message_buf[MAILBOX_MBOX_SIZE];
```
Same for this

glneo commented on drivers/firmware/tisci/tisci.c:
```diff
@@ -0,0 +1,1600 @@
+	struct tisci_xfer xfer;
+	struct tisci_desc desc;
+	struct tisci_version_info version;
+	uint32_t host_id;
```
`host_id` is static for each instance, should go in config. Actually it is already in config also, so no need for this second copy.

Kronosblaster commented on drivers/firmware/tisci/tisci.c:
```diff
@@ -0,0 +1,1599 @@
+	uint32_t host_id;
+};
+
+static struct k_sem response_ready_sem;
```
As of now response_ready_sem cannot be moved to tisci_data since the callback called from the secureproxy driver does not pass user_data  back to context as it was only meant to be a data buffer. This can be added as another commit later.

vaishnavachath commented:
> > @Kronosblaster can we have a test added for different TISCI APIs? so that the client can be verified/tested before integrating to higher level frameworks (power,clock_control .etc)?
> 
> @vaishnavachath Since there is no upstream support for platforms supporting the TISCI driver a test cannot be made as of now for this purpose. A sample application and its output are added to the PR.

Now that the platform support is upstream, can a proper test and the DT changes for tisci node be made part of this PR?

Kronosblaster committed:
drivers: mbox: secureproxy: Pass userdata in isr

This change modifies the secure proxy mailbox driver to pass user data
to the interrupt service routine (ISR) callback. This allows the ISR
to access the specific context or data associated with the mailbox
message being processed.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

natto1784 commented on dts/arm/ti/am64x_main.dtsi:
```diff
@@ -10,6 +10,25 @@
+		interrupts = <0 65 IRQ_TYPE_EDGE IRQ_DEFAULT_PRIORITY>;
+		interrupt-parent = <&vim>;
+		#mbox-cells = <1>;
+		status = "okay";
```
interrupt will be set in `am64x_r5.dtsi` for R5F0_0

glneo reviewed (approved):
There are some minor things I'd still like to cleanup here, but they can be handled in later PRs, this one has had enough nitpicking from me :).

For now this all looks good enough to start.

gramsay0 commented on drivers/firmware/tisci/tisci.c:
```diff
@@ -0,0 +1,1600 @@
+	}
+
+	k_sem_take(&data->data_sem, K_FOREVER);
+	data->seq++;
```
This comment does not appear to be addressed correctly: https://github.com/zephyrproject-rtos/zephyr/pull/90053#discussion_r2126893644

> Since you update seq here, if this function is called again before seq is checked in ti_sci_get_response() then that check will fail to match. This assumed only one message can be in-flight at a time. Which is fine to assume and we do the same in the TF-A/OP-TEE versions of this driver also, but it is not enforced anywhere I can see, so this could be an issue with multi-threaded calls into ti_sci*() functions.

If there is one message in-flight at a time, then a semaphore should be be taken before getting the sequence number and released after the transfer is complete.

The same for modifying `&data->xfer` in this function, another thread could modify it while a transfer is in porgess

gramsay0 replied:
Still not correct, the semaphore should be be taken before getting the sequence number and released after the transfer is complete.

Releasing and retaking the semaphore still allows other threads to preempt

dnltz commented on boards/ti/am243x_evm/am243x_evm_am2434_r5f0_0.dts:
```diff
@@ -56,6 +56,14 @@
+	status = "okay";
+};
+
+&dmsc{
```
```suggestion
&dmsc {
```

dnltz commented on dts/arm/ti/am64x_main.dtsi:
```diff
@@ -10,6 +10,24 @@
 
+	dmsc: system-controller@44043000 {
+		compatible = "ti,k2g-sci";
+		ti,host-id = <36>;
```
Linux says `12` for in `k3-am64x-main.dtsi`

Kronosblaster replied:
[TISCI host IDs](https://software-dl.ti.com/tisci/esd/latest/5_soc_doc/am64x/hosts.html)
Linux runs on A53 cores. This is for the R5 core 0 context 1.

glneo replied:
If this changes for each core, then why is this in the common DTSI file? This number should go in the R5 core 0 specific files. Same for `secure_proxy_main` mboxes and IRQs.

dnltz commented on dts/arm/ti/am64x_main.dtsi:
```diff
@@ -10,6 +10,24 @@
+		compatible = "ti,k2g-sci";
+		ti,host-id = <36>;
+		mbox-names = "rx", "tx";
+		mboxes = <&secure_proxy_main 2>, <&secure_proxy_main 3>;
```
Same here, Linux uses 12 and 13.

Kronosblaster committed:
dts: arm: ti TISCI DMSC support for AM2434 EVM

Device tree edits to add DMSC system controller support for
AM2434 EVM. This includes the addition system-controller and
secure-proxy mailbox nodes in arm46x_main.dtsi.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

Kronosblaster committed:
tests: drivers: tisci: TISCI driver test

TISCI driver test added to validate the TISCI API
functionality in Zephyr. This test is for the target
am243x_evm/am2434/r5f0_0.

Signed-off-by: Dave Joseph <d-joseph@ti.com>

dnltz reviewed (approved):
Please move R5 specific information from `am64_main` to `am64_r5` device-trees, after this PR got merged.

gramsay0 commented on drivers/firmware/tisci/tisci.c:
```diff
@@ -0,0 +1,1602 @@
+{
+	struct rx_msg *msg = user_data;
+
+	k_sem_give(msg->response_ready_sem);
```
I think I see why I'm confused, mbox_ti_secproxy.c dereferences `void *user_data` as a `struct rx_msg *` and does the copy.
https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/mbox/mbox_ti_secproxy.c#L161C3-L161C5

That seems like a bit of a hack, `void *user_data` should always be opaque.

This is well outside the scope of this PR, and I haven't looked at the secure proxy PR, which might explain why it's done this way (I'll take a look and resolve this comment for now)
